## Machine Learning Wine Quality Prediction

In [ ]:
# Import libraries
import pandas as pd
import numpy as np

In [52]:
# Set the working directory to the project folder
import os

os.chdir(r"C:\Users\Kaith\Desktop\WineQT\Datasets")

In [53]:
# Confirm the current working directory and list the files
print(os.getcwd())
print(os.listdir())

C:\Users\Kaith\Desktop\WineQT\Datasets
['WineQT.csv', 'WineQT_Cleaned.csv', 'Wine_Quality_Predictions.csv']


In [12]:
# Load the wine quality prediction dataset
df = pd.read_csv("WineQT_Cleaned.csv")

In [13]:
# Check the dataset demension
df.shape

(1143, 13)

In [14]:
# Display the first five rows
df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,0
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5,1
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5,2
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6,3
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,4


In [15]:
# Check column names and data types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1143 non-null   float64
 1   volatile acidity      1143 non-null   float64
 2   citric acid           1143 non-null   float64
 3   residual sugar        1143 non-null   float64
 4   chlorides             1143 non-null   float64
 5   free sulfur dioxide   1143 non-null   float64
 6   total sulfur dioxide  1143 non-null   float64
 7   density               1143 non-null   float64
 8   pH                    1143 non-null   float64
 9   sulphates             1143 non-null   float64
 10  alcohol               1143 non-null   float64
 11  quality               1143 non-null   int64  
 12  Id                    1143 non-null   int64  
dtypes: float64(11), int64(2)
memory usage: 116.2 KB


In [16]:
# Separete the input features from the target variable
X = df.drop("quality", axis=1)
y = df["quality"]

# Check the shape
print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (1143, 12)
Target shape: (1143,)


In [20]:
# Import train_test_split
from sklearn.model_selection import train_test_split

# Split the database into training and testing sets
X_train, X_test, y_train, y_test = train_test_split (X, y, test_size=0.2, random_state=42)

# Display the shape
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (914, 12)
X_test shape: (229, 12)
y_train shape: (914,)
y_test shape: (229,)


In [23]:
# Import the Linear Regression model
from sklearn.linear_model import LinearRegression

# Create the model
linear_model = LinearRegression()

# Train the model
linear_model.fit(X_train, y_train)

print("Linear Regression model trained successfully!")

Linear Regression model trained successfully!


In [25]:
# Make preedictions using the trained Linear Regression model
y_pred = linear_model.predict(X_test)

# Display the first 10 predictions
print(y_pred[:10])

[5.42239406 4.75039315 5.29684786 5.08885868 6.09486333 6.61357779
 5.4017609  5.12073623 5.80682335 5.32701664]


In [28]:
# Import evaluation metrics
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Calculate evaluation metrics
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

# Display the results
print("Mean Absolute Error (MAE):", mae)
print("Root Mean Squared Error (RMSE):", rmse)
print("R² Score:", r2)

Mean Absolute Error (MAE): 0.4799770439749094
Root Mean Squared Error (RMSE): 0.6184079172594622
R² Score: 0.31276385395083883


In [29]:
# Import Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor

# Create the Random Forest model
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)

# Train the model
rf_model.fit(X_train, y_train)

print("Random Forest model trained successfully!")

Random Forest model trained successfully!


In [43]:
# Make predictions using the Random Forest model
rf_pred = rf_model.predict(X_test)

# Display the first 10 pridictions
print(rf_pred[:10])

[5.28 5.67 5.47 5.2  6.09 6.68 5.18 5.05 5.61 5.06]


In [44]:
# Import evaluation metrics
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Calculate Random Forest evaluation metrics
rf_mae = mean_absolute_error(y_test, rf_pred)
rf_rmse = np.sqrt(mean_squared_error(y_test, rf_pred))
rf_r2 = r2_score(y_test, rf_pred)

# Display the results
print("Random Forest MAE:", rf_mae)
print("Random Forest RMSE:", rf_rmse)
print("Random Forest R² Score:", r2_score(y_test, rf_pred))

Random Forest MAE: 0.4109170305676856
Random Forest RMSE: 0.5458689639136779
Random Forest R² Score: 0.46453299294085415


In [45]:
# Compare model perfpmance

print("Linear Regression")   
print("MAE:", mae) 
print("RMSE:", rmse) 
print("R2 Score:", r2)

print("\nRandom Forest") 
print("MAE:", rf_mae) 
print("RMSE:", rf_rmse) 
print("R2 Score:", rf_r2)       

Linear Regression
MAE: 0.4799770439749094
RMSE: 0.6184079172594622
R2 Score: 0.31276385395083883

Random Forest
MAE: 0.4109170305676856
RMSE: 0.5458689639136779
R2 Score: 0.46453299294085415


In [48]:
# Make final predictions using the Random Forest model
final_predictions = rf_model.predict(X_test)

# Display the first 10 actual and predicted values
prediction_results = pd.DataFrame({"Actual Quality": y_test.values[:10], "Predicted Quality": final_predictions[:10]})
prediction_results.head()

,Actual Quality,Predicted Quality
0,5,5.28
1,6,5.67
2,5,5.47
3,6,5.20
4,6,6.09


In [50]:
# Save the prediction results for Power BI
prediction_results.to_csv("Wine_Quality_Predictions.csv", index=False)

print("Wine quality predictions saved successfully!")

Wine quality predictions saved successfully!


In [40]:
# Calculate the average predicted wine quality
average_predicted_quality = final_predictions.mean()

print("Average Predicted Wine Quality:", average_predicted_quality)

Average Predicted Wine Quality: 5.687685589519651


In [54]:
print("Number of predictions:", len(final_predictions))
print("Average prediction:", final_predictions.mean())

Number of predictions: 229
Average prediction: 5.687685589519651


# Machine Learning — Wine Quality Prediction

## Objective

The objective of this machine learning phase was to develop models capable of predicting wine quality based on the wine characteristics in the cleaned dataset.

## Data Preparation

The dataset was divided into:

* **Features (X):** Wine characteristics used as input variables.
* **Target (y):** `quality`, representing the wine quality score.

The data was split into **80% training data** and **20% testing data** using `train_test_split` with `random_state=42`.

## Models Used

Two regression models were trained and evaluated:

1. **Linear Regression**
2. **Random Forest Regressor**

The models were evaluated using:

* **Mean Absolute Error (MAE)** — measures the average absolute difference between actual and predicted values.
* **Root Mean Squared Error (RMSE)** — measures prediction error while giving greater weight to larger errors.
* **R² Score** — measures how well the model explains variation in wine quality.

## Final Predictions

The Random Forest model was used to generate predictions on the test dataset. The predictions were compared with the actual wine quality values to assess model performance.

The average predicted wine quality was approximately **5.69**.

## Machine Learning Outcome

The machine learning phase successfully produced wine quality predictions and provided a prediction metric that can be incorporated into the Power BI dashboard.

The **Predicted Average Quality KPI** is approximately **5.69**.
